lib/change_stream.js

```diff
@@ -0,0 +1,227 @@
+  self.namespace = {};
+  if (changeDomain.collectionName) {
+    self.namespace.database = self.changeDomain.s.db.databaseName;
+    self.namespace.collection = self.changeDomain.collectionName;
+    self.serverConfig = self.changeDomain.s.db.serverConfig;
+  } else if (self.changeDomain.databaseName) {
+    self.namespace.database = self.changeDomain.databaseName;
+    self.serverConfig = self.changeDomain.serverConfig;
+  } else {
+    throw new Error('changeDomain must be either a database or collection.');
+  }
+
+  self.cursor = createChangeStreamCursor(self);
+
+  self.on('newListener', function(eventName) {
+    if (eventName === 'change') {
+      self.changeListenerExists = true;
+      self.cursor.on('data', function (change) {
+        processNewChange(self, null, change, 'eventEmitter');
+      });
+    }
+  });
+
+};
+
+// Create a new change stream cursor based on self's configuration
+var createChangeStreamCursor = function (self) {
+  var changeStreamCursor;
+
+  if (self.cachedResumeToken) {
+    self.options.resumeAfter = self.cachedResumeToken;
+  }
+
+  var serverSupportsChangeNotifications = false;
+  if (serverSupportsChangeNotifications) {
+    var collectionName = self.changeDomain.collectionName ? self.changeDomain.collectionName : 1;
+    changeStreamCursor = buildChangeStreamAggregationCommand(self, collectionName);
+
+  } else {
+    // Generate a no conflict value for this Change Stream if it does not already exist.
+    // This prevents collisions when multiple change stream simulations exist simulataneously.
+    self.options.cursorNoConflict = self.options.cursorNoConflict || Math.random().toString().substring(2, 12);
+
+    // Create the change stream cursor using the black box simulator
+    changeStreamCursor = changeStreamSimulator.createChangeStreamCursorSimulation(self.serverConfig, self.namespace, self.pipeline, self.options);
+  }
+
+  if (self.changeListenerExists) {
+    changeStreamCursor.on('data', function (change) {
+      processNewChange(self, null, change, 'eventEmitter');
+    });
+  }
+
+  changeStreamCursor.on('close', function() {
+    if (!self.shuttingDown) {
+      self.cursor = createChangeStreamCursor(self);
+    }
+  });
+
+  return changeStreamCursor;
+};
+
+var buildChangeStreamAggregationCommand = function (self, collectionName) {
+    var changeNotificationStageOptions = {};
+    if (self.options.fullDocument) {
+      changeNotificationStageOptions.fullDocument = self.options.fullDocument;
+    }
+    if (self.cachedResumeToken && !self.options.disableResume) {
+      changeNotificationStageOptions.resumeAfter = self.cachedResumeToken;
+    }
+
+    var changeStreamPipeline = [{$changeNotification: changeNotificationStageOptions}].concat(self.pipeline);
+    var command = { aggregate : collectionName, pipeline : changeStreamPipeline };
+
+    // Execute the cursor
+    return self.serverConfig.cursor(self.namespace.database + '.' + collectionName, command, self.options);
```

Comment:
I wonder about this. Isn't the `collectionName` equal to `1` for the database? So you'd have a `return self.serverConfig.cursor(`db.1`, command, self.options);` ?